### PR TITLE
Fix the FSDP.optim_state_dict_to_load OOM

### PR DESCRIPTION
**Pull request by bigning**
# What does this PR do?

Fix the `FSDP.optim_state_dict_to_load` OOM, it's already in pytorch>=2.3.0 https://github.com/pytorch/pytorch/pull/117261

# test
1. **before** the change, it oom in the first forward dbrx-dense-20b-debug-autoresume-AdiVZi

here is the memory before the forward:
image

2. **after** the change, it can train `dbrx-dense-20b-debug-autoresume-3QVblX`
image

### CODE REVIEW

File: composer/trainer/mosaic_fsdp_utils.py [hunk 1]
```diff
@@ -1124,3 +1124,47 @@ def _same_storage(a, b):
...
+        new_optim_state: Dict[str, Any] = {}
+        intra_param_start_idx = shard_param_info.intra_param_start_idx
+        intra_param_end_idx = shard_param_info.intra_param_end_idx
+        for state_name, value in optim_state.items():
+            if (
+                torch.is_tensor(value)
+                and value.dim() > 0
+                and fsdp_state.sharding_strategy != ShardingStrategy.NO_SHARD
+            ):
+                value = value.flatten()[intra_param_start_idx : intra_param_end_idx + 1].clone()  # type: ignore[operator]
```
Comment (snarayan21) [inline composer/trainer/mosaic_fsdp_utils.py:1168]: can you put a comment here indicating that this is the patch, and also link relevant pytorch issues/commits?

**Review by snarayan21 — COMMENTED** [review]
That's a crazy amount of extra memory usage, damn.

Can you make the PR title more descriptive instead of "up"? other than that, one minor comment, otherwise LGTM! thanks for finding and fixing this so quick.

File: composer/trainer/mosaic_fsdp_utils.py [hunk 2]
```diff
@@ -1124,3 +1124,47 @@ def _same_storage(a, b):
...
+
+    @no_type_check
+    def _shard_orig_param_state(
+        fsdp_param_info: FSDPParamInfo,
+        fqn: str,
+        optim_state: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """
+        Shard the optimizer state for the original parameter with the name ``fqn``.
+        This API should only be used when ``use_orig_params`` is True.
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply composer/trainer/mosaic_fsdp_utils.py:1145]: It's patching a [function in torch](https://github.com/pytorch/pytorch/blob/8aa08b8b9d1fab2a13dc5fbda74c553cb2a08729/torch/distributed/fsdp/_optim_utils.py#L376) which doesn't have the full docstring, is that still necessary to add? probably more important to point out what we changed

[the remark this answers is not in the document]
Comment (bigning) [reply composer/trainer/mosaic_fsdp_utils.py:1145]: @karan6181 , this is the just copy of pytorch code. https://github.com/pytorch/pytorch/blob/8aa08b8b9d1fab2a13dc5fbda74c553cb2a08729/torch/distributed/fsdp/_optim_utils.py#L376-L408. I think we usually don't do any thing to this kind of change (even @no_type_check )

**Review by mvpatel2000 — COMMENTED** [review]
Discussed offline, only for 2.2.2 please!

**Review by mvpatel2000 — APPROVED** [review]

**Pull request merged**